libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/viewmodel/MapboxNavigationViewModel.kt

```diff
@@ -0,0 +1,146 @@
+            viewModelScope.launch {
+                _rawLocationUpdates.emit(rawLocation)
+            }
+        }
+        override fun onNewLocationMatcherResult(locationMatcherResult: LocationMatcherResult) {
+            viewModelScope.launch {
+                _newLocationMatcherResult.emit(locationMatcherResult)
+            }
+        }
+    }
+
+    private val routeProgressObserver = RouteProgressObserver {
+        viewModelScope.launch {
+            _routeProgressUpdates.emit(it)
+        }
+    }
+
+    private val routesObserver = RoutesObserver {
+        viewModelScope.launch {
+            _routesUpdatedResults.emit(it)
+        }
+    }
+
+    private val arrivalObserver = object : ArrivalObserver {
+        override fun onFinalDestinationArrival(routeProgress: RouteProgress) {
+            viewModelScope.launch {
+                _finalDestinationArrivals.emit(routeProgress)
+            }
+        }
+
+        override fun onNextRouteLegStart(routeLegProgress: RouteLegProgress) {
+            viewModelScope.launch {
+                _nextRouteLegStartUpdates.emit(routeLegProgress)
+            }
+        }
+
+        override fun onWaypointArrival(routeProgress: RouteProgress) {
+            viewModelScope.launch {
+                _waypointArrivals.emit(routeProgress)
+            }
+        }
+    }
+
+    private val bannerInstructionsObserver = BannerInstructionsObserver { bannerInstructions ->
+        viewModelScope.launch {
+            _bannerInstructions.emit(bannerInstructions)
+        }
+    }
+
+    private val tripSessionStateObserver = TripSessionStateObserver { tripSessionState ->
+        viewModelScope.launch {
+            _tripSessionStateUpdates.emit(tripSessionState)
+        }
+    }
+
+    override fun onStart(owner: LifecycleOwner) {
+        super.onStart(owner)
+        mapboxNavigation.registerLocationObserver(locationObserver)
+        mapboxNavigation.registerRouteProgressObserver(routeProgressObserver)
+        mapboxNavigation.registerRoutesObserver(routesObserver)
+        mapboxNavigation.registerArrivalObserver(arrivalObserver)
+        mapboxNavigation.registerBannerInstructionsObserver(bannerInstructionsObserver)
+        mapboxNavigation.registerTripSessionStateObserver(tripSessionStateObserver)
+    }
+
+    override fun onStop(owner: LifecycleOwner) {
+        super.onStop(owner)
+        mapboxNavigation.unregisterLocationObserver(locationObserver)
+        mapboxNavigation.unregisterRouteProgressObserver(routeProgressObserver)
+        mapboxNavigation.unregisterRoutesObserver(routesObserver)
+        mapboxNavigation.unregisterArrivalObserver(arrivalObserver)
+        mapboxNavigation.unregisterBannerInstructionsObserver(bannerInstructionsObserver)
+        mapboxNavigation.unregisterTripSessionStateObserver(tripSessionStateObserver)
+    }
+
+    override fun onDestroy(owner: LifecycleOwner) {
+        super.onDestroy(owner)
+        mapboxNavigation.onDestroy()
```

Comment:
Since `ViewModel` survives activity recreation (e.g. during configuration change), how are we going to use this destroyed `mapboxNavigation` then?

Reply:
You're right. I moved it to onCleared()